### ISSUE TITLE
Assert failure: executionAborted in `GcInfoDecoder::EnumerateLiveSlots`

### ISSUE THREAD

**Issue by jakobbotsch** [dotnet/runtime#102370]
## Build Information
Build: https://dev.azure.com/dnceng-public/public/_build/results?buildId=678333&view=results
Build error leg or test failing: 
Example console log: https://helixre8s23ayyeko0k025g8.blob.core.windows.net/dotnet-runtime-refs-pull-102261-merge-e822cbb23a0f465186/LibraryImportGenerator.Unit.Tests/1/console.c3aa79dd.log?helixlogtype=result

```
/root/helix/work/workitem/e /root/helix/work/workitem/e
  Discovering: LibraryImportGenerator.Unit.Tests (method display = ClassAndMethod, method display options = None)
  Discovered:  LibraryImportGenerator.Unit.Tests (found 185 of 190 test cases)
  Starting:    LibraryImportGenerator.Unit.Tests (parallel test collections = on [2 threads], stop on fail = off)
    LibraryImportGenerator.UnitTests.Compiles.ValidateSnippetsWithMarshalType [SKIP]
      No current scenarios to test.

Assert failure(PID 28 [0x0000001c], Thread: 73 [0x0049]): executionAborted
    File: /__w/1/s/src/coreclr/vm/gcinfodecoder.cpp:801
    Image: /root/helix/work/correlation/dotnet
```

Maybe related to #101890?

## Error Message

Fill the error message using [step by step known issues guidance](https://github.com/dotnet/arcade/blob/main/Documentation/Projects/Build%20Analysis/KnownIssueJsonStepByStep.md).

```json
{
  "ErrorMessage": "executionAborted",
  "BuildRetry": false,
  "ExcludeConsoleLog": false
}
```

 ### Known issue validation
**Build: :mag_right:** https://dev.azure.com/dnceng-public/public/_build/results?buildId=678333
**Error message validated:** `[executionAborted`]
**Result validation:** :x: Known issue did not match with the provided build.
**Validation performed at:** 5/17/2024 8:39:56 AM UTC

### Report

|Build|Definition|Test|Pull Request|
|---|---|---|---|
|[2503080](https://dev.azure.com/dnceng/internal/_build/results?buildId=2503080)|dotnet-runtime|[LibraryImportGenerator.Unit.Tests.WorkItemExecution](https://dev.azure.com/dnceng//internal/_build/results?buildId=2503080&view=ms.vss-test-web.build-test-results-tab&runId=53931539&resultId=211692)||
|[747517](https://dev.azure.com/dnceng-public/public/_build/results?buildId=747517)|dotnet/runtime|[System.Formats.Tar.Tests.WorkItemExecution](https://dev.azure.com/dnceng-public/public/_build/results?buildId=747517&view=ms.vss-test-web.build-test-results-tab&runId=18899340&resultId=100161)|dotnet/runtime#105099|
|[2490036](https://dev.azure.com/dnceng/internal/_build/results?buildId=2490036)|dotnet-runtime|[System.Text.RegularExpressions.Tests.WorkItemExecution](https://dev.azure.com/dnceng//internal/_build/results?buildId=2490036&view=ms.vss-test-web.build-test-results-tab&runId=53840191&resultId=210843)||
|[2488046](https://dev.azure.com/dnceng/internal/_build/results?buildId=2488046)|dotnet-runtime|[LibraryImportGenerator.Unit.Tests.WorkItemExecution](https://dev.azure.com/dnceng//internal/_build/results?buildId=2488046&view=ms.vss-test-web.build-test-results-tab&runId=53832083&resultId=211477)||
|[2487307](https://dev.azure.com/dnceng/internal/_build/results?buildId=2487307)|dotnet-runtime|[LibraryImportGenerator.Unit.Tests.WorkItemExecution](https://dev.azure.com/dnceng//internal/_build/results?buildId=2487307&view=ms.vss-test-web.build-test-results-tab&runId=53822513&resultId=211085)||
#### Summary
|24-Hour Hit Count|7-Day Hit Count|1-Month Count|
|---|---|---|
|1|1|5|

**Comment by jakobbotsch** [comment]
Also cc @VSadov, I'm not familiar with the logic here, but wonder if it could be related to the new GC safe points.

**Comment by VSadov** [comment]
> Also cc @VSadov, I'm not familiar with the logic here, but wonder if it could be related to the new GC safe points.

With interruptible GC safe points we can stress test each and every safe point. 
In the past we would skip a good portion of safe points - JIT helpers, direct calls, for example would not be stress tested.  

So - the change to safe points could be involved here, but it could also expose some existing bug.

Does this happen without the PR change?

**Comment by jakobbotsch** [comment]
> Does this happen without the PR change?

I don't know, I haven't investigated.
Happened again in the recent libraries-jitstress run:
Pipeline run: https://dev.azure.com/dnceng-public/public/_build/results?buildId=690710&view=results
Job name: net9.0-linux-Release-arm-disabler2r
Console log: https://helixre8s23ayyeko0k025g8.blob.core.windows.net/dotnet-runtime-refs-heads-main-705beed46fcf4015ad/LibraryImportGenerator.Unit.Tests/1/console.f4051e9b.log?helixlogtype=result

**Comment by jakobbotsch** [comment]
Saw one on linux-x64 as well in https://dev.azure.com/dnceng-public/public/_build/results?buildId=692730&view=results.
Console log: https://helixre107v0xdcypoyl9e7f.blob.core.windows.net/dotnet-runtime-refs-pull-102808-merge-ca406cfc9a8448aea7/LibraryImportGenerator.Unit.Tests/1/console.b56c6cd1.log?helixlogtype=result

**Comment by davidwrighton** [comment]
Happened again on linux-arm https://dev.azure.com/dnceng-public/public/_build/results?buildId=698579&view=ms.vss-test-web.build-test-results-tab&runId=17382094&resultId=208376&paneView=dotnet-dnceng.dnceng-build-release-tasks.helix-test-information-tab

Console log
[https://helixre8s23ayyeko0k025g8.blob.core.windows.net/dotnet-runtime-refs-pull-101580-merge-24599464203f4fedbf/LibraryImportGenerator.Unit.Tests/1/console.4e5f55dd.log?helixlogtype=result](https://helixre8s23ayyeko0k025g8.blob.core.windows.net/dotnet-runtime-refs-pull-101580-merge-24599464203f4fedbf/LibraryImportGenerator.Unit.Tests/1/console.4e5f55dd.log?helixlogtype=result)

**Comment by VSadov** [comment]
Always on arm32 and always in  `LibraryImportGenerator.Unit.Tests`

The failure is strange though. It means that we try to initiate a stack walk in a fully-interruptible method, but the IP happens to not be in one of the interruptible ranges. It is hard to think of how this could happen as anything that leads to stack walks should at some point would ask "is this IP interruptible?".

**Comment by VSadov** [comment]
The last failure is interesting as it is not in a JIT stress run.

**Comment by jakobbotsch** [comment]
> The failure is strange though. It means that we try to initiate a stack walk in a fully-interruptible method, but the IP happens to not be in one of the interruptible ranges. It is hard to think of how this could happen as anything that leads to stack walks should at some point ask "is this IP interruptible?".

Here is 100% reliable repro on linux-x64: https://gist.github.com/jakobbotsch/b7b98e082d7be7a189d2edc23b375a89

I think the issue was exposed by #101761. It is similar in spirit to #102919, #103300, #104042, where the JIT is incorrectly handling the fact that we now have a managed helper call in a new location.
In this particular case the tailcall logic turns off GC before it starts placing arguments to the tailcall in the incoming parameters area (the GC ness of those args may not match the parameters, hence we need to do that). However, there is a block copy calling into the bulk write barrier helper after the JIT disables GC.
So I think the underlying problem is that the JIT turns off GC too early. We need similar reasoning as added by #103301. 

I am looking into a fix.

**Comment by VSadov** [comment]
@jakobbotsch - Also the first instruction of an epilog should not have GC forbidden. It was fixed in https://github.com/dotnet/runtime/pull/104336/files#r1664973906 , but right now reverted. 

In theory that also could cause `Assert failure: executionAborted` - if we need to start stackwalk in a leaf method after a call and epilog starts right after that call as well.

**Comment by AndyAyersMS** [comment]
> @jakobbotsch - Also the first instruction of an epilog should not have GC forbidden. It was fixed in https://github.com/dotnet/runtime/pull/104336/files#r1664973906 , but right now reverted.
> 
> In theory that also could cause `Assert failure: executionAborted` - if we need to start stackwalk in a leaf method after a call and epilog starts right after that call as well.

Maybe you can PR that separately, if the fix for the reversion will take a while?

**Comment by VSadov** [comment]
> Maybe you can PR that separately, if the fix for the reversion will take a while?

I am considering that, but need to be sure it was not the part of the PR that made JIT stress unhappy.

**Comment by VSadov** [comment]
The partial re-revert, that contains only JIT fixes has some GC stress failures (extra ones compared to a no-op GC stress off main).  Suspiciously the failures are all on arm64 and with JITStress. 

So, it is possible that it is the JIT changes for the No-GC range in the epilog that make JIT stress unhappy. 

Maybe the "fix" was incomplete and there is one more thing, perhaps in in JIT stress itself, that indirectly relies on old behavior.

**Comment by VSadov** [comment]
> I am considering that, but need to be sure it was not the part of the PR that made JIT stress unhappy.

Turns out it was the "fix" that exposed bad GC info on the instruction right after `CORINFO_HELP_FAIL_FAST`. It was always possible to have it incorrect, but as long as we treated the first instruction of epilog as uninterruptible, it was mitigating the issue.

Treating the first instruction of epilog as uninterruptible is not the right thing to do, in general, especially after calls that could return or could do GC.  
`CORINFO_HELP_FAIL_FAST` is fairly special in that sense - it does not return and does not do GC, but also the way we emit it may result in corrupted GC info on the next instruction. 

I have a few solutions in mind, but need to figure what actually works more naturally.  
Re: https://github.com/dotnet/runtime/pull/105596#issuecomment-2259013925
